Extend the BREAK mode of Timer 1 and 8

rolandvs opened the pull request:
When using `PWM` we have a `deadtime` parameter that is used to prevent a short when using the two channel PWM to drive a bridge. Along this function the BRK_IN function is also very handy to kill the PWM drive when something goes wrong (and a bit earlier than the software can tell you :-)

rolandvs committed:
Extend the BREAK mode of Timer 1 and 8

dpgeorge commented:
Thanks for the patch, it's a good feature to enable.

I would suggest that it be configured via a single new keyword argument rather than two.  There would be 3 states: disabled, enabled and low polarity, enabled and high polarity (eg similar to UART parity).  Since `break` is a keyword the option could be named `break_mode`.  Keep in mind that such a feature could eventually be available on other ports.

rolandvs committed:
Added brk modes

rolandvs committed:
Set the printed values in UC

rolandvs committed:
Remove the brk_pol parameter and combine with brk_mode

Remove the keyword
Check the brk_mode keyword and split them between two register bits

rolandvs committed:
Changed description of brk_mode function

rolandvs committed:
Changed init function and default value

rolandvs committed:
Remove the brk_polarity keyword argument

rolandvs committed:
Add three constants for the brk_mode

dpgeorge commented:
Thanks for updating the PR.

To shorten the keyword and constants, how about `brk` for the keyword and `BRK_OFF`, `BRK_LOW`, `BRK_HIGH` for the constants?  Eg: `timer.init(brk=timer.BRK_LOW)`.  That's still readable enough.

rolandvs committed:
shorten the const names

rolandvs committed:
print the correct keywords

rolandvs commented:
Squash into one?

dpgeorge commented on ports/stm32/timer.c:
```diff
@@ -573,6 +591,7 @@ STATIC mp_obj_t pyb_timer_init_helper(pyb_timer_obj_t *self, size_t n_args, cons
         { MP_QSTR_div,          MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = 1} },
         { MP_QSTR_callback,     MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_rom_obj = MP_ROM_PTR(&mp_const_none_obj)} },
         { MP_QSTR_deadtime,     MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = 0} },
+        { MP_QSTR_break_mode,   MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = BRK_OFF} },
```
I'd suggest also to change this keyword arg to `brk` to match the constants and keep it short.

dpgeorge commented:
> Squash into one?

I can do that during merge.

rolandvs committed:
For consistency break -> brk

dpgeorge commented:
Thanks for updating.  I know I suggested `break_mode` initially, but what I meant above was to use simply `brk`, not `brk_mode` (eg similar to Pin's `pull`, it's not `pull_mode`).  I think it's obvious from the context that you are setting a mode so that word doesn't need to be part of the keyword, `brk` is enough and keeps it short.

I'm happy to make this change when merging.

rolandvs commented:
As they say: "be my guest" :-)

dpgeorge commented:
Updated and merged in edd0e0f93d180e79c1422d1eb7a9ec1954bb6d03

chrismas9 commented:
@rolandvs I am working on an STM33L4plus (L4P5 & L4Rx) port and want to test the Timers. I can't work out how to use the `brk` feature. There does not appear to be any way to route a `BKIN` pin through to the Timer block. There is no call to `HAL_TIMEx_ConfigBreakInput` which sets the `brk` source on the `OR2 register`. I can't set an `af` on an input pin to forward a `BKIN pin` to the Timer block.

How can your `brk` feature be used and tested? Can a PWM channel be shutdown by a `BKIN pin`?

Thanks,
Chris

rolandvs commented:
Dear Chris,

Yes the BKIN is for shutting down the PWM immediately. It has been a while and I need to hook up a board and get a code sample I made back then. But currently I am on holidays and did not pack a laptop and a few uPy boards for fun with me...

I do remember that I used the STM32F091RC and configured some pins like:
```
# pwm signal and inverse, with dead-band and fault input
pin_CH1  = Pin(Pin.cpu.A8, mode=Pin.AF_PP, af=Pin.AF2_TIM1)
pin_CH1N = Pin(Pin.cpu.A7, mode=Pin.AF_PP, af=Pin.AF2_TIM1)
pin_BKIN = Pin(Pin.cpu.A6, mode=Pin.AF_PP, af=Pin.AF2_TIM1)
```

Setting a timer was something like:

```
def timerOn(val, dt=0):
    global ne555
    global ne555_ch1
    ne555.init(freq=val+FREQ_OFFSET, deadtime=dt)
    ne555_ch1 = ne555.channel(1, Timer.PWM, pulse_width=(ne555.period()+1)//2, pin=Pin.cpu.A8)
```

All I can do is set up a PWM on Timer1 and check the BKIN function, but you need to wait for another week... :-)

chrismas9 commented:
Hi Roland, thanks for your prompt reply and the code example. It's interesting that you set the BKIN to AF.PP which is an output mode. You have given me a good starting point to experiment. Enjoy your holiday!

Regards,
Chris

rolandvs commented:
Oops, that's not how it should be. The pin should be in `AF2_TIM1` mode. The `BKIN` is purely hardware so the pin definition is not necessary. The two extra parameters need to be included after the `deadtime=0` named variable.

Tnx and when I return I will test it again... I need to look around if I have a L4R/L4P board...

chrismas9 commented:
I have an F091 board. I will test on that first until I have working code before trying it on L4+. There are other issues with L4+ like missing af definitions in the HAL. I don't want to be debugging two things at once.

chrismas9 commented:
Hi Roland,

`pin_BKIN = Pin(Pin.cpu.A6, mode=Pin.AF_PP, af=Pin.AF2_TIM1)` works, although it's not intuitive and not documented in [https://github.com/micropython/micropython/pull/6535](url)
`pin_BKIN = Pin(Pin.cpu.A6, mode=Pin.IN, af=Pin.AF2_TIM1)` doesn't work, even though you want the pinas in input.

You can only set `alt `for `Pin.AF_PP` and `Pin.AF_OD` modes. This sets the mode bits to 0b10 and connects the pin to the AF mux. It is disconnected from the GPIO diver so the output mode is ignored and the pin stays as an input for BKIN.

I have decided to push my L4+ work as is, then have a good look at timers and fix the issues for L4+ separately. I would also like to add external clk_in on ETR as I use that and have to poke the registers to enable it. You can't do that on the L0 and F0 because `stm.mem32 `is not available.

Regards, Chris

chrismas9 commented:
`machine.Pin` is more logical. `mode=machine.Pin.ALT` does not imply a direction. This works:
`pin_BKIN = machine.Pin(machine.Pin.cpu.A6, mode=machine.Pin.ALT, alt=machine.Pin.AF2_TIM1)`

chrismas9 commented:
Due to the common code in `pin.c` for `pyb.Pin` amd `machine.Pin` the following works for `pyb.Pin`:
`pin_BKIN = Pin(Pin.cpu.A6, mode=Pin.ALT, alt=Pin.AF2_TIM1)`
Using `mode=Pin.ALT` instead of `mode=Pin.AF_PP` for an input is much more logical.
I will update the docs.
